### ISSUE TITLE
Help required?

### ISSUE THREAD

**Issue by rowanwins** [mfogel/polygon-clipping#5]
Gday @mfogel 

Nice work on this so far. Is there anything in particular you'd like help with to mature this repo?

 - I might submit a benchmark test tonight to compare this against @w8r 's version of martinez? I think this is a good starting point to.
- Are you happy with the test suite?
- If the above are looking good would you like me to put together a demo page perhaps?

Cheers
Rowan

**Comment by rowanwins** [comment]
Gday @mfogel 

So I've just run a benchmark test of this against @w8r 's version and so far this is significantly slower although I've not really done any digging as to why that might be.

````
Martinez x 7.77 ops/sec ±3.83% (24 runs sampled)
polygon-clipping x 0.44 ops/sec ±1.23% (6 runs sampled)
````

I haven't bothered to fork things yet and so I cant easily push up my branch but if you're interested let me know.

I think you mentioned elsewhere that you a re less concerned about performance and more concerned with stability so Im not sure how much effort you'll put into speeding things up. It's a discussion worth having because I think performance is important to @w8r and myself.

Looks great though, super readable code!

**Comment by mfogel** [comment]
Awesome @rowanwins , thank you much in advance :)

If you can put up a branch or PR with the benchmark code set up, that would be awesome. I would like to get in there and do some profiling and understand why the drop in performance is that much. I would have guessed more like 2-4x, but that's more like 20x. I'm hopeful there will be some low hanging fruit to get that more reasonable.

I am happy with the test suite. I think we'd generally be ok to just add to it as bugs are found and fixed.

A demo page would be helpful! I imagine you have some ideas from that. I actually never saw the demo page for martinez - [the live one](http://w8r.name/martinez/demo/) errors out in my browser and I never ran it locally. Whatever it had would probably be a good starting point.

Not necessary for a first version of a demo page, but a feature I was thinking would be nice would be to provide some way for people to easily see the test cases covered in the [end-to-end portion](https://github.com/mfogel/polygon-clipping/tree/v0.5.0/test/end-to-end) of the testsuite. So like either a quick way to load each one, or maybe just a page full of iframes or divs of the github-rendered geojson files.

**Comment by rowanwins** [comment]
Righto well I'll try and take a look into the performance issues over the next few nights and report back.

**Comment by mfogel** [comment]
Hi @rowanwins I found time today to work on performance optimizations. Using the benchmark for binary ops from martinez, here's what I get with what's on master now for the Asia union (I think that's the result you have above there?)

```
Martinez x 7.52 ops/sec ±6.61% (23 runs sampled)
polygon-clipping x 1.95 ops/sec ±6.79% (9 runs sampled)
```

Given that this library supports:

 * overlapping multipolygons
 * N input multipolygons - not just two

I don't expect there to be much additional room for performance improvements without some pretty invasive changes. But words like that are said to be proven wrong ;)

I think it would be good to add to the benchmark some cases with N input multipolygons - 10, 100, 1000. It would be interesting to see at what N polygon-clipping begins to run faster than martinez or jsts (assuming it does - I have not verified that yet!).

**Comment by rowanwins** [comment]
Gday @mfogel 

Sorry for not getting back to this. 

On my initial inspection I did see lots of use of `.forEach` loops which I gather are usually slower than traditional `for` loops, there are a fair few articles online about this eg [here](https://stackoverflow.com/questions/43821759/why-array-foreach-is-slower-than-for-loop-in-javascript). So it may be that you could get some performance gains out of switching those loops although how much of a difference that would make in the scheme of things I'm not sure.

And yeah you are right about having some benchmarks for various vertice lengths (10,100, 1000 etc), we haven't done this yet (the asia scenario is very large, but the rest of our test cases are very small).

**Comment by mfogel** [comment]
Thanks for the `forEach` tip... I'm still relatively new to the javascript world. I'll do a pass switching those out where it makes sense.

**Comment by mfogel** [comment]
For future reference, I ran some benchmarks on my own machine (node v9.8.0, mac) comparing the different types of for lops I know about in JS:

```
Trivial for loops to sum an array of length 10
for (let i = 0; i < values.length; i++) {}                   x 48,783,231 ops/sec ±0.94% (88 runs sampled)
for (let i = 0, length = values.length; i < length; i++) {}  x 47,685,397 ops/sec ±1.01% (84 runs sampled)
for (var i = 0, length = values.length; i < length; i++) {}  x 47,508,453 ops/sec ±1.55% (88 runs sampled)
values.forEach()                                             x 4,685,729 ops/sec ±2.07% (90 runs sampled)
for (const val of values) {}                                 x 26,597,539 ops/sec ±1.03% (89 runs sampled)
for (let val of values) {}                                   x 27,583,572 ops/sec ±1.77% (81 runs sampled)
for (var val of values) {}                                   x 27,262,110 ops/sec ±1.24% (88 runs sampled)
```

```
Trivial for loops to sum an array of length 10000
for (let i = 0; i < values.length; i++) {}                   x 77,739 ops/sec ±0.95% (91 runs sampled)
for (let i = 0, length = values.length; i < length; i++) {}  x 77,405 ops/sec ±1.44% (93 runs sampled)
for (var i = 0, length = values.length; i < length; i++) {}  x 77,641 ops/sec ±0.91% (91 runs sampled)
values.forEach()                                             x 7,984 ops/sec ±1.04% (93 runs sampled)
for (const val of values) {}                                 x 39,656 ops/sec ±1.27% (91 runs sampled)
for (let val of values) {}                                   x 40,591 ops/sec ±0.99% (93 runs sampled)
for (var val of values) {}                                   x 40,605 ops/sec ±1.03% (93 runs sampled)
```

```
Trivial for loops to sum an array of length 10000000
for (let i = 0; i < values.length; i++) {}                   x 68.09 ops/sec ±4.09% (67 runs sampled)
for (let i = 0, length = values.length; i < length; i++) {}  x 59.69 ops/sec ±4.96% (62 runs sampled)
for (var i = 0, length = values.length; i < length; i++) {}  x 59.40 ops/sec ±3.03% (61 runs sampled)
values.forEach()                                             x 4.24 ops/sec ±8.77% (15 runs sampled)
for (const val of values) {}                                 x 34.09 ops/sec ±6.75% (47 runs sampled)
for (let val of values) {}                                   x 36.95 ops/sec ±2.80% (49 runs sampled)
for (var val of values) {}                                   x 36.63 ops/sec ±2.62% (49 runs sampled)
```

Takeaways:

 * traditional for loop is fastest
 * `for (const i of vaules)` is maybe 60% as fast as traditional for loop
 * `forEach` is 10x slower (probably also applies to use of `map`, `some`, `every`... etc)

It appears that there isn't really a difference between:

 * `const`/`let` vs. `var`
 * assigning an intermediate variable like `length = values.length`

**Comment by rowanwins** [comment]
Good summary thanks @mfogel , it will be interesting to see what sort of impact that has across the board in your codebase. 

The only other thing I spotted would be the use of `.filter`, I'd hazard a guess and say this could be performed quicker using a more hand-written function, it's in a similar basket to `.forEach` where it is convenient but has some overhead.

**Comment by mfogel** [comment]
Ok, so after doing another pass to remove most all of those functional array-walking methods (`forEach`, `filter`, `map`, `some` ... )

```
Martinez x 9.95 ops/sec ±4.81% (29 runs sampled)
polygon-clipping x 4.52 ops/sec ±2.64% (15 runs sampled)
```

Now within a factor of, well, almost 2. Better than I dared to hope a week ago... 10x improvement since then! I think code readability and maintainability has dropped a bit in the performance-optimization process, but such is life.

@rowanwins if you see any other similar red flags, so to speak, in the code - please let me know. That was obviously super helpful.

**Comment by rowanwins** [comment]
Wow good result!

With your loops you can cache the length of whatever you're looping, another minor thing but will help a bit given that this algorithm is quite intensive on loops.
````
var len = values.length
for (let i = 0; i < len; i++) {}   
```

Also for your event queue you have an isEmpty check which may be a bit slow. Currently you have
````
  get isEmpty () {
    return this.tinyQueue.length === 0
  }
````
but I think it's faster to check for an empty array by doing
````
  get isEmpty () {
    return this.tinyQueue.length
  }
````
Again this is a micro-optimization but given that it's being called a squillion time it might help.

**Comment by rowanwins** [comment]
I've noticed a bunch more changes, how are the latest benchmarks tracking?

**Comment by mfogel** [comment]
Benchmarks are a bit better with the latest round of changes, now hovering around 50% of martinez on the Asia union:

```
Martinez x 10.08 ops/sec ±5.53% (30 runs sampled)
polygon-clipping x 5.06 ops/sec ±2.80% (19 runs sampled)
```

The asia union is a good test for me because it is pretty similar to how I'm using this package in my data pipeline.

I'm going to write up another ticket (later today?) that details out a performance shortcut I want to do. The primary purpose will beto reduce memory usage (memory usage is an issue in my data pipeline) but a side effect is that for operations like the asia union, it should improve performance as well. I should be able to get that sketched out today.

**Comment by erf** [comment]
You are probably aware of this already but i thought i mention a few things. I think, as you mention, reducing memory usage will help performance. I see you use a lot of object allocation ( new SomeObject ) inside various loops. Maybe you could move some of those objects out of the loops and reuse them, or maybe avoid creating them altogether ( to make the GC do less work ). I'm not sure how the various compilers ( V8 ) do memory / handle object creation, but i would try to make use of the Stack instead of Heap where possible, and reuse objects on the Heap where possible. Usually primitives are handled by the stack, and objects on the Heap. I'm not sure, but i think object literals performs better than constructor functions - a quick benchmark i did showed a 40% performance gain in favor of object literals.

**Comment by mfogel** [comment]
@erf thanks for the ideas. I should be able to find time this weekend to try using object literals instead of constructor functions.

Once place where I'm confident a performance gain could be found - right now for every line segment in a polygon, 3 objects are created - two `SweepEvent` and one `Segment`. The `Segment` object could be merged with one of those `SweepEvent` objects to just create two objects - a `LeftSweepEvent` and a `RightSweepEvent`. I haven't done this because it's a lot of moving code around (and making it harder to understand) and really may only save a few percent.

**Comment by erf** [comment]
@mfogel i don't understand all the code but would it make sense to move the `cache Map` out of the `getLeftmostComparator` function of `sweep-event.js` so you don't create that many instances of it and maybe do more caching?

**Comment by mfogel** [comment]
@erf following up on your last comment, I just did a simple benchmark comparing those two and also get the object is significantly more performant than the array. http://jsben.ch/T3Aeh

I'll change point implementation in this lib from array to object literal today. Stay tuned...

**Comment by erf** [comment]
@mfogel i ran the benchmarks with various results dependent on the browser - in Chrome and Safari code block 1 was significantly faster ( 8-12x ), but in Firefox code block 2 was ~5x faster. So all in all it seem block 1 might be a better choice.

**Comment by mfogel** [comment]
In the commit above basically all point operations are changed from `[x, y]` to `{x: x, y: y}`. According to the [benchmarks in martinez](https://github.com/w8r/martinez/blob/v0.3.3/bench.js), it appears to give a slight performance boost - around 5-10%. (I was hoping for more, but at least it was a step in the right direction)

**Comment by erf** [comment]
It is at least more readable imo. Btw the reasoning for not using Arrays is because Arrays usually have to do bound checking to see if the index is out of scope.

**Comment by erf** [comment]
I tried to do some profiling using the v8-profiler, but i was not able to analyze the results in the chrome browser. Maybe i’ll figure it out later..

**Comment by erf** [comment]
I would like to port this library to Swift and maybe other languages but i see `geom-in` and `geom-out` both have classes with the same name, which makes it difficult to port without naming conflicts. Maybe the classes could be simplified to just use function calls like in the original implementation, or combined into other classes, and maybe put into separate files. It would perhaps make the implementation more clearer also.

**Comment by erf** [comment]
I did some benchmarking and found out that removing the following got me a 3x performance boost using the `hole_hole.geojson` set.
```
    cleanInput.forceMultiPoly(geoms[i])
    cleanInput.cleanMultiPoly(geoms[i])
```

**Comment by faroscore** [comment]
@mfogel Hello! Thank you very much for your work, but how I should use the lib with ES5? Splaytree in the bundle has ES6 class declaration.

**Comment by rowanwins** [comment]
Hi @faroscore 

What makes you think you can't use it with ES5, are you getting any errors? 
When it's published the src directory is compiled using webpack and babel to be ES5 friendly [dist directory](https://github.com/mfogel/polygon-clipping/tree/master/dist) so you shouldn't be running into issues (depending on how you're using it).

**Comment by faroscore** [comment]
> Hi @faroscore
> 
> What makes you think you can't use it with ES5, are you getting any errors?
> When it's published the src directory is compiled using webpack and babel to be ES5 friendly [dist directory](https://github.com/mfogel/polygon-clipping/tree/master/dist) so you shouldn't be running into issues (depending on how you're using it).

https://github.com/mfogel/polygon-clipping/blob/master/dist/polygon-clipping.js#L107 Look at this line, eval contains class keyword. I got syntax error in IE11.

**Comment by mfogel** [comment]
I think what's going on with the failure in ie11 is that babel isn't compiling [this file](https://github.com/w8r/splay-tree/blob/master/index.js) that polygon-clipping is importing from splaytree. That file has the class declarations that are showing up in `dist/polygon-clipping.js`

I am far from an expert in all the lovely world of javascript build/compile/package etc tools, but it sounds like from this [blog post](https://babeljs.io/blog/2018/06/26/on-consuming-and-publishing-es2015+-packages#how-babel-v7-helps) that upgrading to babel 7 might fix this, so I'll give that a try tomorrow.

**Comment by mfogel** [comment]
I think I fixed the ES5 and ie11 issues without upgrading babel. There were two things going on:

 * the imports from Splaytree - these also needed to be transpiled by babel because polygon-clipping is importing an ES6 module.
 * use of `.includes()` on arrays isn't transformed by default for some reason by babel, but I found a babel plugin to do it.

In any case the library now appears to work for me on ie11.

(Note: I didn't commit re-built files for the `dist/` directory yet, so to test this out in your own dev env the first step would be to run `npm run build:lib` to re-build those files)

I'm planning to do a new release of this lib to NPM within the week, that release will include this fix.
